feat(basic create) add submitItemCreate

yurkov-anton opened the pull request:
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add the basic function for the "Create" load operations

Signed-off-by: Anton Yurkov ayurkov@gmail.com

yurkov-anton committed:
feat(basic creat) add submitItemCreate

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add the basic function for the "Create" load operations

Signed-off-by: Anton Yurkov ayurkov@gmail.com

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +167,50 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+
+            System.out.format("Writing inBuffer: '%s' to stream '%s / %s'%n",
+                    inBuff, scope, streamName);
+            final CompletableFuture writeFuture = writer.writeEvent(inBuff);
```
* await the event is written (the returned future is done) before method exits
* update the event task status (status code, done bytes count)

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +167,50 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+                                                                                 null,
+                                                                                 EventWriterConfig.builder().build())) {
+            final String routingKey = null;
+            final ByteBuffer inBuff = DirectMemUtil.getThreadLocalReusableBuff(eventSize);
```
replace this with the correct data buffer
```
final DataInput dataInput = eventItem.dataInput();
final ByteBuff outBuff = dataInput.layer(eventItem.layer());
// TODO set the correct offset, etc
```

yurkov-anton committed:
issue-2-itemCreate

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add future hadling.
Add the event task status updating.
Fix buffer creation.

Signed-off-by: Anton Yurkov ayurkov@gmail.com

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +167,65 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+            final DataInput dataInput = eventItem.dataInput();
+            final ByteBuffer outBuff = dataInput.getLayer(eventItem.layer());
+
+            final CompletableFuture writeFuture = writer.writeEvent(outBuff);
```
need to set the offset to the output buffer before
the offset may be determined by calling eventItem.offset() method
also, it's necessary to set the output buffer limit to make it writing no more than the eventItem.size() bytes

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +167,65 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+
+            final CompletableFuture writeFuture = writer.writeEvent(outBuff);
+            try {
+                writeFuture.get();
```
writeFuture is a CompleatableFuture instance
the result should be handled using:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#handle-java.util.function.BiFunction-

yurkov-anton committed:
fix future handling, add offset setting
By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anton Yurkov ayurkov@gmail.com

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +168,72 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+                    .asReadOnlyBuffer();
+
+            outBuff.position((int)offset);
+            outBuff.limit(dataSize);
```
Will throw an `IllegalArgumentException` if the `dataSize` value is more than the remaining byte count in the buffer. The limit value should be a minimum of two:
- dataSize
- buffer remaining bytes

Let's save that minimum into the new `int n` local variable and use it instead of the `dataSize` further in this method.

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +168,72 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+            return false;
+        };
+
+        StreamConfiguration streamConfig = StreamConfiguration.builder()
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +168,72 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+        streamManager.createStream(scope, streamName, streamConfig);
+
+        try (ClientFactory clientFactory = ClientFactory.withScope(scope, URI.create(uriSchema));
+             EventStreamWriter<ByteBuffer> writer = clientFactory.createEventWriter(streamName,
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -163,6 +168,72 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+             EventStreamWriter<ByteBuffer> writer = clientFactory.createEventWriter(streamName,
+                     null,
+                     EventWriterConfig.builder().build())) {
+            final String routingKey = null;
```
what is this line for?

veronikaKochugova commented:
please resolve the conflict

yurkov-anton committed:
add сheck "if the volume remaining in the buffer"

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anton Yurkov ayurkov@gmail.com

yurkov-anton committed:
update this branch to master
Merge branch 'master' into issue-2-itemCreate

yurkov-anton committed:
change name sabmitItemCreate to invokeItemCreate

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

change name sabmitItemCreate to invokeItemCreate
add create streemManager

Signed-off-by: Anton Yurkov ayurkov@gmail.com

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -360,6 +237,73 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+			return false;
+		};
+
+		StreamConfiguration streamConfig = StreamConfiguration.builder()
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -360,6 +237,73 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+				.scalingPolicy(ScalingPolicy.fixed(1))
+				.build();
+
+        StreamManager streamManager = StreamManager.create(URI.create(uriSchema));
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -360,6 +237,73 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+        StreamManager streamManager = StreamManager.create(URI.create(uriSchema));
+		streamManager.createStream(scope, streamName, streamConfig);
+
+		try (ClientFactory clientFactory = ClientFactory.withScope(scope, URI.create(uriSchema));
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -360,6 +237,73 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+		streamManager.createStream(scope, streamName, streamConfig);
+
+		try (ClientFactory clientFactory = ClientFactory.withScope(scope, URI.create(uriSchema));
+			 EventStreamWriter<ByteBuffer> writer = clientFactory.createEventWriter(streamName,
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -227,6 +237,73 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+
+			try {
+				writeFuture.get();
+			} catch ( Exception e) {
```
final

veronikaKochugova commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -227,6 +237,73 @@ private void invokeDirectoryNio(PathOperation<? extends PathItem> pathOperation)
+			final long buffSize;
+			try {
+				buffSize = eventItem.size();
+			} catch (IOException e) {
```
final
